### PR TITLE
[#3199] Allow gateways to omit tenant ID with HTTP adapter

### PR DESCRIPTION
**Pull request by sophokles73**
This is the first step for #3199

The HTTP protocol adapter has been changed to allow authenticated
gateway devices to omit the tenant ID from the URI of requests
used to upload telemetry or event messages.

### CODE REVIEW

File: adapters/http-vertx-base/src/test/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapterTest.java [hunk 1]
```diff
@@ -840,18 +865,22 @@ public void testMessageLimitExceededForACommandResponseMessage() {
...
 
-    private HttpContext newHttpContext(final Buffer payload, final HttpServerResponse response) {
-        return newHttpContext(payload, mock(HttpServerRequest.class), response);
+    private HttpServerRequest newEventRequest() {
+        final HttpServerRequest request = mock(HttpServerRequest.class);
+        when(request.uri()).thenReturn("/" + EventConstants.EVENT_ENDPOINT);
+        return request;
     }
 
     private HttpContext newHttpContext(
```
Comment (calohmn) [inline adapters/http-vertx-base/src/test/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapterTest.java:886]: Method not used any more.

File: adapters/sigfox-vertx/src/main/java/org/eclipse/hono/adapter/sigfox/impl/SigfoxProtocolAdapter.java [hunk 2]
```diff
@@ -159,20 +159,21 @@ private void dataHandler(final HttpContext ctx, final UploadHandler uploadHandle
...
         if (!requestTenant.equals(deviceTenant)) {
-            ctx.fail(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST,
+            ctx.fail(new ClientErrorException(HttpURLConnection.HTTP_FORBIDDEN,
                     "tenant information mismatch"));
             return;
         }
 
-        final String contentType = (data != null) ? CONTENT_TYPE_OCTET_STREAM
-                : EventConstants.CONTENT_TYPE_EMPTY_NOTIFICATION;
+        final String contentType = Optional.ofNullable(data)
```
Comment (calohmn) [inline adapters/sigfox-vertx/src/main/java/org/eclipse/hono/adapter/sigfox/impl/SigfoxProtocolAdapter.java:174]: `data` can't be `null` here.

File: adapters/http-vertx-base/src/test/java/org/eclipse/hono/adapter/http/impl/VertxBasedHttpProtocolAdapterTest.java [hunk 3]
```diff
@@ -365,6 +409,40 @@ public void testPutEventSucceedsForValidCredentials(final VertxTestContext ctx)
                 }));
     }
 
+    /**
+     * Verifies that a request to upload eventsusing PUT fails if the request contains a malformed URI.
```
Comment (calohmn) [inline adapters/http-vertx-base/src/test/java/org/eclipse/hono/adapter/http/impl/VertxBasedHttpProtocolAdapterTest.java:413]: typo ("eventsusing")

File: adapter-base/src/main/java/org/eclipse/hono/adapter/HttpContext.java [hunk 4]
```diff
@@ -241,28 +255,25 @@ public Device getAuthenticatedDevice() {
...
-
-            if (timeTilDisconnectHeader.isPresent()) {
-                return Integer.parseInt(timeTilDisconnectHeader.get());
-            }
-        } catch (final NumberFormatException e) {
-        }
-
-        return null;
+        return Optional.ofNullable(request().getHeader(Constants.HEADER_TIME_TILL_DISCONNECT))
+                .or(() -> Optional.ofNullable(request().getParam(Constants.HEADER_TIME_TILL_DISCONNECT)))
```
[the remark this answers is not in the document]
Comment (sophokles73) [reply adapter-base/src/main/java/org/eclipse/hono/adapter/HttpContext.java:268]: I am not sure if I understand your point. Can you elaborate?

[the remark this answers is not in the document]
Comment (calohmn) [reply adapter-base/src/main/java/org/eclipse/hono/adapter/HttpContext.java:268]: The HTTP adapter documentation only mentions a `hono-ttd` *request header*:
https://www.eclipse.org/hono/docs/user-guide/http-adapter/#publish-an-event-authenticated-device

It isn't mentioned that also a `hono-ttd` *query parameter* can be used.

[the remark this answers is not in the document]
Comment (sophokles73) [reply adapter-base/src/main/java/org/eclipse/hono/adapter/HttpContext.java:268]: Ah, now I got it. I Will change that accordingly ...

**Comment by sophokles73** [comment]
@calohmn I have pushed changes

**Review by calohmn — APPROVED** [review]
LGTM

**Pull request merged**